File: thorvg.flutter/android/src/main/kotlin/com/robotoss/thorvg_plus/ThorvgPlusPlugin.kt
package com.robotoss.thorvg_plus

import android.os.Handler
import android.os.HandlerThread
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.view.TextureRegistry

/**
 * Method-channel host for the Texture-based renderer.
 *
 * **Single shared render thread**: thorvg's lock implementation is a no-op
 * unless `TaskScheduler::threads() > 0`, and we initialize it with 0 worker
 * threads. To keep cross-instance global state (LoaderMgr, Initializer
 * refcount) safe, every JNI call across every texture runs on this one
 * `HandlerThread`. The Flutter UI isolate is still freed; we trade
 * intra-screen parallelism for correctness.
 *
 * The legacy `dart:ffi` path that drives `Lottie` via `CustomPaint` is
 * unaffected and remains usable in parallel.
 */
class ThorvgPlusPlugin : FlutterPlugin, MethodCallHandler {

    companion object {
        const val CHANNEL = "thorvg_plus/texture"

        init {
            // Resolves at class-load time so a missing or mis-packaged native
            // binary surfaces as a clear `UnsatisfiedLinkError` immediately
            // rather than as an opaque JNI failure on the first method call.
            System.loadLibrary("thorvg")
        }
    }

    private var channel: MethodChannel? = null
    private var textureRegistry: TextureRegistry? = null
    private val textures = HashMap<Long, ThorvgTexture>()

    private var renderThread: HandlerThread? = null
    private var renderHandler: Handler? = null

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(binding.binaryMessenger, CHANNEL).apply {
            setMethodCallHandler(this@ThorvgPlusPlugin)
        }
        textureRegistry = binding.textureRegistry

        val t = HandlerThread("thorvg-render").apply { start() }
        renderThread = t
        renderHandler = Handler(t.looper)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        synchronized(textures) {
            textures.values.toList().forEach { runCatching { it.dispose() } }
            textures.clear()
        }
        channel?.setMethodCallHandler(null)
        channel = null
        textureRegistry = null

        // Drain any in-flight render/dispose tasks before stopping the
        // looper, so we don't strand a thorvg destroy call.
        renderHandler = null
        renderThread?.quitSafely()
        renderThread = null
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "create" -> handleCreate(call, result)
            "play" -> handleSimple(call, result) { it.play() }
            "pause" -> handleSimple(call, result) { it.pause() }
            "dispose" -> handleDispose(call, result)
            "seek" -> handleSeek(call, result)
            "resize" -> handleResize(call, result)
            else -> result.notImplemented()
        }
    }

    private fun handleCreate(call: MethodCall, result: Result) {
        val registry = textureRegistry
        val handler = renderHandler
        if (registry == null || handler == null) {
            result.error("NO_REGISTRY", "Plugin not attached", null)
            return
        }
        val data = call.argument<ByteArray>("data")
        val width = call.argument<Int>("width") ?: 0
        val height = call.argument<Int>("height") ?: 0
        val animate = call.argument<Boolean>("animate") ?: true
        val repeat = call.argument<Boolean>("repeat") ?: true
        val reverse = call.argument<Boolean>("reverse") ?: false
        val speed = call.argument<Double>("speed") ?: 1.0

        if (data == null || data.isEmpty()) {
            result.error("BAD_ARGS", "missing 'data'", null)
            return
        }
        if (width <= 0 || height <= 0) {
            result.error("BAD_ARGS", "width/height must be > 0 (got ${width}x${height})", null)
            return
        }

        ThorvgTexture.createAsync(
            registry = registry,
            handler = handler,
            data = data,
            width = width,
            height = height,
            animate = animate,
            repeat = repeat,
            reverse = reverse,
            speed = speed,
        ) { texResult ->
            texResult.fold(
                onSuccess = { tex ->
                    synchronized(textures) { textures[tex.id] = tex }
                    result.success(
                        mapOf(
                            "textureId" to tex.id,
                            "lottieWidth" to tex.lottieWidth,
                            "lottieHeight" to tex.lottieHeight,
                            "totalFrame" to tex.totalFrame.toDouble(),
                            "duration" to tex.duration.toDouble(),
                        )
                    )
                },
                onFailure = { e ->
                    result.error("CREATE_FAILED", e.message ?: e.toString(), null)
                }
            )
        }
    }

    private inline fun handleSimple(
        call: MethodCall,
        result: Result,
        action: (ThorvgTexture) -> Unit,
    ) {
        val id = (call.arguments as? Map<*, *>)?.get("textureId") as? Number
        if (id == null) {
            result.error("BAD_ARGS", "missing 'textureId'", null)
            return
        }
        val tex = synchronized(textures) { textures[id.toLong()] }
        if (tex == null) {
            result.error("NOT_FOUND", "no texture for id $id", null)
            return
        }
        runCatching { action(tex) }
            .onSuccess { result.success(null) }
            .onFailure { result.error("CALL_FAILED", it.message, null) }
    }

    private fun handleDispose(call: MethodCall, result: Result) {
        val id = (call.arguments as? Map<*, *>)?.get("textureId") as? Number
        if (id == null) {
            result.error("BAD_ARGS", "missing 'textureId'", null)
            return
        }
        val tex = synchronized(textures) { textures.remove(id.toLong()) }
        if (tex == null) {
            // already disposed — treat as success
            result.success(null)
            return
        }
        runCatching { tex.dispose() }
            .onSuccess { result.success(null) }
            .onFailure { result.error("DISPOSE_FAILED", it.message, null) }
    }

    private fun handleSeek(call: MethodCall, result: Result) {
        val id = call.argument<Number>("textureId")
        val frame = call.argument<Number>("frame")
        if (id == null || frame == null) {
            result.error("BAD_ARGS", "need 'textureId' and 'frame'", null)
            return
        }
        val tex = synchronized(textures) { textures[id.toLong()] }
        if (tex == null) {
            result.error("NOT_FOUND", "no texture for id $id", null)
            return
        }
        tex.seek(frame.toFloat())
        result.success(null)
    }

    private fun handleResize(call: MethodCall, result: Result) {
        val id = call.argument<Number>("textureId")
        val w = call.argument<Number>("width")
        val h = call.argument<Number>("height")
        if (id == null || w == null || h == null) {
            result.error("BAD_ARGS", "need 'textureId', 'width', 'height'", null)
            return
        }
        val tex = synchronized(textures) { textures[id.toLong()] }
        if (tex == null) {
            result.error("NOT_FOUND", "no texture for id $id", null)
            return
        }
        tex.resize(w.toInt(), h.toInt())
        result.success(null)
    }
}

File: thorvg.flutter/ios/Classes/ThorvgPlusPlugin.swift
import Flutter
import UIKit

/// Method-channel host for the Texture-based renderer on iOS.
///
/// Mirrors the Android `ThorvgPlusPlugin`:
///   - one shared `MethodChannel` per engine,
///   - a registry of `ThorvgTexture` instances keyed by `textureId`,
///   - each texture owns its own `DispatchQueue` and `CVPixelBufferPool`,
///   - rasterization happens on the texture's queue, never on the main isolate.
///
/// The legacy `dart:ffi` path is unchanged and continues to work in parallel.
public class ThorvgPlusPlugin: NSObject, FlutterPlugin {

    private static let channelName = "thorvg_plus/texture"

    private let textureRegistry: FlutterTextureRegistry
    private var textures: [Int64: ThorvgTexture] = [:]
    private let texturesLock = NSLock()

    /// Shared render queue. thorvg's internal locks are no-ops while
    /// `TaskScheduler::threads() == 0`, so cross-instance global state
    /// (LoaderMgr, Initializer refcount, …) must be reached from one
    /// serial queue. Per-texture parallelism is sacrificed; the UI thread
    /// stays free, which is the primary goal.
    private let renderQueue = DispatchQueue(
        label: "io.thorvg.render", qos: .userInitiated
    )

    init(registry: FlutterTextureRegistry) {
        self.textureRegistry = registry
        super.init()
    }

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(
            name: channelName,
            binaryMessenger: registrar.messenger()
        )
        let instance = ThorvgPlusPlugin(registry: registrar.textures())
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "create":  handleCreate(call, result: result)
        case "play":    handleSimple(call, result: result) { $0.play() }
        case "pause":   handleSimple(call, result: result) { $0.pause() }
        case "dispose": handleDispose(call, result: result)
        case "seek":    handleSeek(call, result: result)
        case "resize":  handleResize(call, result: result)
        default:        result(FlutterMethodNotImplemented)
        }
    }

    private func handleCreate(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let dataAny = args["data"],
              let width = (args["width"] as? NSNumber)?.intValue,
              let height = (args["height"] as? NSNumber)?.intValue
        else {
            result(FlutterError(code: "BAD_ARGS", message: "missing data/width/height", details: nil))
            return
        }
        guard width > 0, height > 0 else {
            result(FlutterError(code: "BAD_ARGS", message: "width/height must be > 0", details: nil))
            return
        }

        let bytes: Data
        if let typed = dataAny as? FlutterStandardTypedData {
            bytes = typed.data
        } else if let raw = dataAny as? Data {
            bytes = raw
        } else {
            result(FlutterError(code: "BAD_ARGS", message: "'data' must be Uint8List", details: nil))
            return
        }

        let animate = (args["animate"] as? NSNumber)?.boolValue ?? true
        let repeats = (args["repeat"] as? NSNumber)?.boolValue ?? true
        let reverse = (args["reverse"] as? NSNumber)?.boolValue ?? false
        let speed = (args["speed"] as? NSNumber)?.doubleValue ?? 1.0

        // Run all native init (Initializer::init / SwCanvas::gen / load) on
        // the plugin's shared render queue. Reply over MethodChannel from
        // there — FlutterResult is thread-safe.
        ThorvgTexture.createAsync(
            registry: textureRegistry,
            queue: renderQueue,
            data: bytes,
            width: width,
            height: height,
            animate: animate,
            repeats: repeats,
            reverse: reverse,
            speed: speed
        ) { [weak self] texResult in
            switch texResult {
            case .success(let texture):
                self?.texturesLock.lock()
                self?.textures[texture.textureId] = texture
                self?.texturesLock.unlock()
                result([
                    "textureId": NSNumber(value: texture.textureId),
                    "lottieWidth": NSNumber(value: texture.lottieWidth),
                    "lottieHeight": NSNumber(value: texture.lottieHeight),
                    "totalFrame": NSNumber(value: Double(texture.totalFrame)),
                    "duration": NSNumber(value: Double(texture.duration)),
                ])
            case .failure(let error):
                result(FlutterError(code: "CREATE_FAILED", message: "\(error)", details: nil))
            }
        }
    }

    private func handleSimple(
        _ call: FlutterMethodCall,
        result: @escaping FlutterResult,
        action: (ThorvgTexture) -> Void
    ) {
        guard let id = textureId(from: call) else {
            result(FlutterError(code: "BAD_ARGS", message: "missing 'textureId'", details: nil))
            return
        }
        guard let texture = lookup(id) else {
            result(FlutterError(code: "NOT_FOUND", message: "no texture for id \(id)", details: nil))
            return
        }
        action(texture)
        result(nil)
    }

    private func handleDispose(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let id = textureId(from: call) else {
            result(FlutterError(code: "BAD_ARGS", message: "missing 'textureId'", details: nil))
            return
        }
        texturesLock.lock()
        let texture = textures.removeValue(forKey: id)
        texturesLock.unlock()
        texture?.dispose()
        result(nil)
    }

    private func handleSeek(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let id = (args["textureId"] as? NSNumber)?.int64Value,
              let frame = (args["frame"] as? NSNumber)?.floatValue
        else {
            result(FlutterError(code: "BAD_ARGS", message: "need 'textureId' and 'frame'", details: nil))
            return
        }
        guard let texture = lookup(id) else {
            result(FlutterError(code: "NOT_FOUND", message: "no texture for id \(id)", details: nil))
            return
        }
        texture.seek(frame: frame)
        result(nil)
    }

    private func handleResize(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let id = (args["textureId"] as? NSNumber)?.int64Value,
              let w = (args["width"] as? NSNumber)?.intValue,
              let h = (args["height"] as? NSNumber)?.intValue
        else {
            result(FlutterError(code: "BAD_ARGS", message: "need textureId/width/height", details: nil))
            return
        }
        guard let texture = lookup(id) else {
            result(FlutterError(code: "NOT_FOUND", message: "no texture for id \(id)", details: nil))
            return
        }
        texture.resize(width: w, height: h)
        result(nil)
    }

    private func textureId(from call: FlutterMethodCall) -> Int64? {
        guard let args = call.arguments as? [String: Any] else { return nil }
        return (args["textureId"] as? NSNumber)?.int64Value
    }

    private func lookup(_ id: Int64) -> ThorvgTexture? {
        texturesLock.lock(); defer { texturesLock.unlock() }
        return textures[id]
    }
}
